API not starting

I downloaded the latest code from Git, built  it and started API.
API shuts down with these errors:

Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.security.ldap.authentication.ad.ActiveDirectoryLdapAuthenticationProvider]: Factory method 'activeDirectoryLdapAuthenticationProvider' threw exception; nested exception is java.lang.IllegalArgumentException: Url cannot be empty
        at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:189)
        at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:588)
        ... 23 more
Caused by: java.lang.IllegalArgumentException: Url cannot be empty
        at org.springframework.util.Assert.isTrue(Assert.java:68)
        at org.springframework.security.ldap.authentication.ad.ActiveDirectoryLdapAuthenticationProvider.<init>(ActiveDirectoryLdapAuthenticationProvider.java:118)
        at com.capitalone.dashboard.config.WebSecurityConfig.activeDirectoryLdapAuthenticationProvider(WebSecurityConfig.java:49)
        at com.capitalone.dashboard.config.WebSecurityConfig$$EnhancerBySpringCGLIB$$7e96236b.CGLIB$activeDirectoryLdapAuthenticationProvider$2(<generated>)
        at com.capitalone.dashboard.config.WebSecurityConfig$$EnhancerBySpringCGLIB$$7e96236b$$FastClassBySpringCGLIB$$d829ef8a.invoke(<generated>)
        at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:228)
        at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:355)
        at com.capitalone.dashboard.config.WebSecurityConfig$$EnhancerBySpringCGLIB$$7e96236b.activeDirectoryLdapAuthenticationProvider(<generated>)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:162)


This is the configuration file:
# Database Name
dbname=dashboarddb

# Database HostName - default is localhost
dbhost=localhost

# Database Port - default is 27017
dbport=27017

# MongoDB replicaset
#dbreplicaset=[false if you are not using MongoDB replicaset]
#dbhostport=[host1:port1,host2:port2,host3:port3]

# Database Username - default is blank
dbusername=dashboarduser

# Database Password - default is blank
dbpassword=dbpassword

server.contextPath=/api
server.port=8080

This was done based on your comments in #1329 

Can you advise what could be wrong here?

--- a/api/README.md
+++ b/api/README.md
@@ -50,6 +50,8 @@ auth.ldapUserDnPattern=[uid={0},OU=Users,dc=your,dc=company,dc=com]
 auth.adDomain=[company.com]
 # This will be your root dn
 auth.adRootDn=[dc=your,dc=company,dc=com]
+# This will be your active directory url (required for AD)
+auth.adUrl=[Need an example]
 
 monitor.proxy.host=[hostname of proxy server]
 monitor.proxy.type=[http|socks|direct]

--- a/api/docker/properties-builder.sh
+++ b/api/docker/properties-builder.sh
@@ -50,6 +50,8 @@ auth.ldapUserDnPattern=${AUTH_LDAP_USER_DN_PATTERN:-}
 auth.adDomain=${AUTH_AD_DOMAIN:-}
 # This will be your root dn
 auth.adRootDn=${AUTH_AD_ROOT_DN:-}
+# This is your active directory url
+auth.adUrl=${AUTH_AD_URL:-}
 
 #Monitor Widget proxy credentials
 monitor.proxy.username=${MONITOR_PROXY_USERNAME:-}

--- a/api/src/main/java/com/capitalone/dashboard/auth/AuthProperties.java
+++ b/api/src/main/java/com/capitalone/dashboard/auth/AuthProperties.java
@@ -23,6 +23,7 @@ public class AuthProperties {
 	private String ldapServerUrl;
 	private String adDomain;
 	private String adRootDn;
+	private String adUrl;
 	
 	public void setExpirationTime(Long expirationTime) {
 		this.expirationTime = expirationTime;
@@ -71,6 +72,14 @@ public class AuthProperties {
 	public void setAdRootDn(String adRootDn) {
 		this.adRootDn = adRootDn;
 	}
+	
+    public String getAdUrl() {
+        return adUrl;
+    }
+
+    public void setAdUrl(String adUrl) {
+        this.adUrl = adUrl;
+    }
 
 	@PostConstruct
 	public void applyDefaultsIfNeeded() {

--- a/api/src/main/java/com/capitalone/dashboard/config/WebSecurityConfig.java
+++ b/api/src/main/java/com/capitalone/dashboard/config/WebSecurityConfig.java
@@ -1,6 +1,5 @@
 package com.capitalone.dashboard.config;
 
-import com.capitalone.dashboard.auth.ldap.CustomUserDetailsContextMapper;
 import org.apache.commons.lang3.StringUtils;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.boot.autoconfigure.security.Http401AuthenticationEntryPoint;
@@ -19,6 +18,7 @@ import org.springframework.security.web.authentication.UsernamePasswordAuthentic
 
 import com.capitalone.dashboard.auth.AuthProperties;
 import com.capitalone.dashboard.auth.AuthenticationResultHandler;
+import com.capitalone.dashboard.auth.ldap.CustomUserDetailsContextMapper;
 import com.capitalone.dashboard.auth.ldap.LdapLoginRequestFilter;
 import com.capitalone.dashboard.auth.standard.StandardLoginRequestFilter;
 import com.capitalone.dashboard.auth.token.JwtAuthenticationFilter;
@@ -40,18 +40,6 @@ public class WebSecurityConfig extends WebSecurityConfigurerAdapter {
 	
 	@Autowired
 	private AuthProperties authProperties;
-
-	@Bean
-	protected ActiveDirectoryLdapAuthenticationProvider activeDirectoryLdapAuthenticationProvider() {
-		ActiveDirectoryLdapAuthenticationProvider provider =
-				new ActiveDirectoryLdapAuthenticationProvider(authProperties.getAdDomain(),
-						authProperties.getLdapServerUrl(),
-						authProperties.getAdRootDn());
-		provider.setConvertSubErrorCodesToExceptions(true);
-		provider.setUseAuthenticationRequestCredentials(true);
-		provider.setUserDetailsContextMapper(new CustomUserDetailsContextMapper());
-		return provider;
-	}
 	
 	@Override
 	protected void configure(HttpSecurity http) throws Exception {
@@ -81,19 +69,25 @@ public class WebSecurityConfig extends WebSecurityConfigurerAdapter {
     @Override
 	protected void configure(AuthenticationManagerBuilder auth) throws Exception {
 		auth.authenticationProvider(standardAuthenticationProvider);
+		configureLdap(auth);
+		configureActiveDirectory(auth);
 
-		if (StringUtils.isEmpty(authProperties.getAdDomain())) {
-			String ldapServerUrl = authProperties.getLdapServerUrl();
-			String ldapUserDnPattern = authProperties.getLdapUserDnPattern();
-			if (StringUtils.isNotBlank(ldapServerUrl) && StringUtils.isNotBlank(ldapUserDnPattern)) {
-				auth.ldapAuthentication()
-				.userDnPatterns(ldapUserDnPattern)
-				.contextSource().url(ldapServerUrl);
-			}
-		} else {
-			auth.authenticationProvider(activeDirectoryLdapAuthenticationProvider());
-		}
 	}
+
+    private void configureActiveDirectory(AuthenticationManagerBuilder auth) {
+        ActiveDirectoryLdapAuthenticationProvider adProvider = activeDirectoryLdapAuthenticationProvider();
+        if(adProvider != null) auth.authenticationProvider(adProvider);
+    }
+
+    private void configureLdap(AuthenticationManagerBuilder auth) throws Exception {
+        String ldapServerUrl = authProperties.getLdapServerUrl();
+		String ldapUserDnPattern = authProperties.getLdapUserDnPattern();
+		if (StringUtils.isNotBlank(ldapServerUrl) && StringUtils.isNotBlank(ldapUserDnPattern)) {
+			auth.ldapAuthentication()
+			.userDnPatterns(ldapUserDnPattern)
+			.contextSource().url(ldapServerUrl);
+		}
+    }
 	
 	@Bean
 	protected StandardLoginRequestFilter standardLoginRequestFilter() throws Exception {
@@ -104,5 +98,17 @@ public class WebSecurityConfig extends WebSecurityConfigurerAdapter {
 	protected LdapLoginRequestFilter ldapLoginRequestFilter() throws Exception {
 		return new LdapLoginRequestFilter("/login/ldap", authenticationManager(), authenticationResultHandler);
 	}
+	
+    @Bean
+    protected ActiveDirectoryLdapAuthenticationProvider activeDirectoryLdapAuthenticationProvider() {
+        if(StringUtils.isBlank(authProperties.getAdUrl())) return null;
+        
+        ActiveDirectoryLdapAuthenticationProvider provider = new ActiveDirectoryLdapAuthenticationProvider(authProperties.getAdDomain(), authProperties.getAdUrl(),
+                authProperties.getAdRootDn());
+        provider.setConvertSubErrorCodesToExceptions(true);
+        provider.setUseAuthenticationRequestCredentials(true);
+        provider.setUserDetailsContextMapper(new CustomUserDetailsContextMapper());
+        return provider;
+    }
 	
 }

